[quant][fx] Make scale, zero_point buffers in the model and use FQN (for quantized ops)

supriyar opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #51259 [quant][fx] Update name of packed weight attributes
* #51171 [quant][fx] Make scale, zero_point buffers in the model, use FQN (for quantize_per_tensor ops)
* **#51166 [quant][fx] Make scale, zero_point buffers in the model and use FQN (for quantized ops)**
* #51086 [quant][fx] Scope support for call_function in QuantizationTracer

Summary:

Currently scale and zero_point values are stored as constant values in the graph.
This prevents these values from being updated in the graph and also does not enable saving
these values to state_dict

After this PR we store scale/zero_point values for quantized ops as buffers in the root module
and create get_attr nodes for them in the graph.

NOTE: The scale and zero_point values will now be stored as part of the model state_dict of the FX quantized model.
Before this PR, we were not storing any qparams in state_dict so saving/loading a quantized fx model didn't work.

We also use the FQN of the module where the quantized ops are present to name these attributes so
that they can be uniquely  identified and mapped to quantized ops.

Test Plan:
python test/test_quantization.py TestQuantizeFx.test_qparams_buffers

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D26092965](https://our.internmc.facebook.com/intern/diff/D26092965)

supriyar committed:
[quant][fx] Make scale, zero_point buffers in the model and use FQN (for quantized ops)
Summary:

Currently scale and zero_point values are stored as constant values in the graph.
This prevents these values from being updated in the graph and also does not enable saving
these values to state_dict

After this PR we store scale/zero_point values for quantized ops as buffers in the root module
and createe get_attr nodes for them in the graph.

We also use the FQN of the module where the quantized ops are present to name these attributes so
that they can be uniquely  identified and mapped to quantized ops.

Test Plan:
python test/test_quantization.py TestQuantizeFx.test_qparams_buffers

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 commented on test/quantization/test_quantize_fx.py:
```diff
@@ -1495,6 +1495,58 @@ def test_convtranspose_per_channel_fails_early(self):
+
+        scale_count = 0
+        zero_point_count = 0
+        for k in keys:
```
should we verify that the key contains qualified path of the module as well?

supriyar replied:
Sure, I can add that

vkuzo commented:
does this change the state dict of existing models?  Are there any BC things to worry about?

jerryzh168 reviewed (approved):
looks good, had one minor comment about test

supriyar commented:
> does this change the state dict of existing models? Are there any BC things to worry about?

@vkuzo before this we had an empty state_dict for fx quantized models that had functional ops. This PR will add values to the state_dict. 
@jerryzh168 Do we have a set of models that we should be testing against?

vkuzo commented:
> > does this change the state dict of existing models? Are there any BC things to worry about?
> 
> @vkuzo before this we had an empty state_dict for fx quantized models that had functional ops. This PR will add values to the state_dict.
> @jerryzh168 Do we have a set of models that we should be testing against?

makes sense, thanks.  So, should we add handling for making sure loading a state_dict from before this PR works correctly with strict==True?  I might be missing something, but seems like that would fail as is since now there would be new keys that are expected?

supriyar committed:
Update on "[quant][fx] Make scale, zero_point buffers in the model and use FQN (for quantized ops)"


Summary:

Currently scale and zero_point values are stored as constant values in the graph.
This prevents these values from being updated in the graph and also does not enable saving
these values to state_dict

After this PR we store scale/zero_point values for quantized ops as buffers in the root module
and createe get_attr nodes for them in the graph.

We also use the FQN of the module where the quantized ops are present to name these attributes so
that they can be uniquely  identified and mapped to quantized ops.

Test Plan:
python test/test_quantization.py TestQuantizeFx.test_qparams_buffers

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D26092965](https://our.internmc.facebook.com/intern/diff/D26092965)

[ghstack-poisoned]

supriyar committed:
Update on "[quant][fx] Make scale, zero_point buffers in the model and use FQN (for quantized ops)"


Summary:

Currently scale and zero_point values are stored as constant values in the graph.
This prevents these values from being updated in the graph and also does not enable saving
these values to state_dict

After this PR we store scale/zero_point values for quantized ops as buffers in the root module
and create get_attr nodes for them in the graph.

NOTE: The scale and zero_point values will now be stored as part of the model state_dict of the FX quantized model.
Before this PR, we were not storing any qparams in state_dict so saving/loading a quantized fx model didn't work.

We also use the FQN of the module where the quantized ops are present to name these attributes so
that they can be uniquely  identified and mapped to quantized ops.

Test Plan:
python test/test_quantization.py TestQuantizeFx.test_qparams_buffers

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D26092965](https://our.internmc.facebook.com/intern/diff/D26092965)

[ghstack-poisoned]

supriyar committed:
Update on "[quant][fx] Make scale, zero_point buffers in the model and use FQN (for quantized ops)"


Summary:

Currently scale and zero_point values are stored as constant values in the graph.
This prevents these values from being updated in the graph and also does not enable saving
these values to state_dict

After this PR we store scale/zero_point values for quantized ops as buffers in the root module
and create get_attr nodes for them in the graph.

NOTE: The scale and zero_point values will now be stored as part of the model state_dict of the FX quantized model.
Before this PR, we were not storing any qparams in state_dict so saving/loading a quantized fx model didn't work.

We also use the FQN of the module where the quantized ops are present to name these attributes so
that they can be uniquely  identified and mapped to quantized ops.

Test Plan:
python test/test_quantization.py TestQuantizeFx.test_qparams_buffers

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D26092965](https://our.internmc.facebook.com/intern/diff/D26092965)

[ghstack-poisoned]

supriyar committed:
Update on "[quant][fx] Make scale, zero_point buffers in the model and use FQN (for quantized ops)"


Summary:

Currently scale and zero_point values are stored as constant values in the graph.
This prevents these values from being updated in the graph and also does not enable saving
these values to state_dict

After this PR we store scale/zero_point values for quantized ops as buffers in the root module
and create get_attr nodes for them in the graph.

NOTE: The scale and zero_point values will now be stored as part of the model state_dict of the FX quantized model.
Before this PR, we were not storing any qparams in state_dict so saving/loading a quantized fx model didn't work.

We also use the FQN of the module where the quantized ops are present to name these attributes so
that they can be uniquely  identified and mapped to quantized ops.

Test Plan:
python test/test_quantization.py TestQuantizeFx.test_qparams_buffers

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D26092965](https://our.internmc.facebook.com/intern/diff/D26092965)

[ghstack-poisoned]

supriyar committed:
Update on "[quant][fx] Make scale, zero_point buffers in the model and use FQN (for quantized ops)"


Summary:

Currently scale and zero_point values are stored as constant values in the graph.
This prevents these values from being updated in the graph and also does not enable saving
these values to state_dict

After this PR we store scale/zero_point values for quantized ops as buffers in the root module
and create get_attr nodes for them in the graph.

NOTE: The scale and zero_point values will now be stored as part of the model state_dict of the FX quantized model.
Before this PR, we were not storing any qparams in state_dict so saving/loading a quantized fx model didn't work.

We also use the FQN of the module where the quantized ops are present to name these attributes so
that they can be uniquely  identified and mapped to quantized ops.

Test Plan:
python test/test_quantization.py TestQuantizeFx.test_qparams_buffers

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D26092965](https://our.internmc.facebook.com/intern/diff/D26092965)

[ghstack-poisoned]
